File: Leetcode/algorithms/easy/0001_two_sum/0001_two_sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

class Solution {
public:
	std::vector<int> twoSum(std::vector<int> &nums, int target) {

		if (nums.size() == 2)
			return {0, 1};

		std::unordered_map<int, int> hMap;

		for (int i = 0; i < nums.size(); i++) {
			if (hMap.find(nums[i]) != hMap.end())
				return {hMap[nums[i]], i};
			else
				hMap[target - nums[i]] = i;
		}

		return {};
	}
};

int main() {
	Solution sol;
	std::vector<int> nums = {2, 7, 11, 15};
	int target = 9;
	std::vector<int> result = sol.twoSum(nums, target);
	for (int i = 0; i < result.size(); i++)
	{
		std::cout << result[i] << " ";
	}
	std::cout << std::endl;
	return 0;
}

File: Leetcode/algorithms/easy/0001_two_sum/0001_two_sum.go
func twoSum(nums []int, target int) []int {
    // Base Case
    if len(nums) == 2 {
        return []int{0, 1}
    }

    // Create a map to store the indices of the numbers that have been seen so far
    hMap := make(map[int]int)

    for i, num := range nums {
        // Check if the complement of num has been seen before
        complement := target - num
        if j, ok := hMap[complement]; ok {
            return []int{j, i}
        }
        // Add num and its index to the map
        hMap[num] = i
    }

    return []int{}
}

// Example usage
func main() {
    nums := []int{2, 7, 11, 15}
    target := 9
    result := twoSum(nums, target)
    fmt.Println(result) // prints [0 1]
}
